### ISSUE TITLE
Accessibility features for screen readers

### ISSUE THREAD

**Issue by foliovision** [foliovision/freedom-video-player#693]
Original issue link: [#995 Accessibility features for screen readers](https://github.com/flowplayer/flowplayer/issues/995)

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Feb 7, 2016

See: https://flowplayer.org/forum/#!/setup:flowplayer-6-screen-reader

Problems:
- testing requires a screen reader installed
- according to experiences with Flowplayer Flash there's basically only one program around - JAWS http://www.freedomscientific.com/Products/Blindness/JAWS - Used to work with IE only, now seems to support Firefox

@nnarhinen - we should come to a decision whether we are _planning_ to implement this or not.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/10995666?v=4' /> [deadplant](https://github.com/deadplant) commented on Feb 7, 2016

for testing there is a free plugin screen-reader for chome https://chrome.google.com/webstore/detail/chromevox/kgejglhpjiefppelpmljglcjbhoiplfn?hl=en

It should be a relatively simple feature to implement when using the html5 engine.  At a minimum, allowing the user to tab focus through the player controls and providing good alt-text for them should be enough for the screen reader.

It (accessibility including screen reading) is becoming a requirement for more and more governments and organizations.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/265438?v=4' /> [nnarhinen](https://github.com/nnarhinen) commented on Feb 7, 2016

I'm all for it, no timeframe though

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/10995666?v=4' /> [deadplant](https://github.com/deadplant) commented on Feb 23, 2016

Here is some more info that I hope might help put someone on the right track when/if they go to work on this.

There are essentially 3 things we need for the screen reader to work well.
- tab-key navigation of the player controls.   (a tabindex on each relevant div or anchor)
- for DIVs a title attribute on each of those.   title="volume slider" so the reader can speak something sensible
- for anchor tags you need to use the anchor text area, not the title attribute for the speech-text.   a common technique is to add a span inside the anchor and style it to position it -1000px off screen.  (making it hidden using visibility causes issues apparently)
- keyboard activation of the focused player control.  enter/spacebar for most things, perhaps arrow keys for the volume slider and scrub bar.

pitfalls:
- focused elements should get a css outline when they have been given focus by the keyboard (this is the desired behaviour for accessibility)  _but_ should not get the outline when given focus by a mouse click.
- screen readers screw with click and keydown events!
  http://unobfuscated.blogspot.ca/2013/05/event-handlers-and-screen-readers.html
  For DIV based controls for example, you will need to add the role="menuitem" attribute if you want to receive keydown events.

hopefully that is helpful to somebody at some point.  :)

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Aug 30, 2016

I wonder whether something could be done with the builtin keyboard shortcut support: https://flowplayer.org/docs/setup.html#keyboard

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/10995666?v=4' /> [deadplant](https://github.com/deadplant) commented on Aug 30, 2016

tabbing in and through the player controls (in the html5 player engine) simply requires a tabindex property on each control's html element.

As a temporary solution I am adding those properties after the player finishes rendering.
Then, to activate the selected control via spacebar/enter I bind a callback to the element like this:
$('.fp-mute').keypress(function(evt){ if (evt.keyCode == 13 || evt.keyCode == 32) { player.mute(); } });

If the built-in keyboard controls could catch space/enter keystrokes and activate whichever tabindex has focus that would be nice.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/5694569?v=4' /> [qualia100](https://github.com/qualia100) commented on Aug 30, 2016

The builtin keyboard shortcuts are great; the problem is - you can't access them without a mouse. Some people are unable to use a mouse and don't even have it. This is why it is important to be able to use a tab to 'activate' those controls. Right now, only mouseover (mouseenter, mouseleave) activate but there is a need to add focus to the that group. Of course, I can adjust the code myself; however I would prefer not to as javascript is not my area of expertise. 

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/5694569?v=4' /> [qualia100](https://github.com/qualia100) commented on Sep 20, 2016

In order to be 508 compliant, it should be possible to use a player with a keyboard only (no mouse). Currently, it is not possible as users need to 'activate' the player area in order to get keyboard shortcuts (https://flowplayer.org/docs/setup.html#keyboard) and that can't be done without a mouse. New functionality that would allow users to use tab to select player area would be very useful. Ideally, users should be able to tab through all flowplayer's controls and activate them with a keyboard. 
Please notify when this functionality could be added to the player.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/5694569?v=4' /> [qualia100](https://github.com/qualia100) commented on Sep 23, 2016

Fullscreen keyboard access ('f' when mouse is over player's area) does not work in Edge. Works fine in other browsers.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/168338?v=4' /> [phloxic](https://github.com/phloxic) commented on Sep 24, 2016

Also does not work in Mac OS Safari. Never did.

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/5694569?v=4' /> [qualia100](https://github.com/qualia100) commented on Feb 21, 2017

My understanding is this is planned for 7.1 release. Is there a timeframe for that release? The ability to access and activate player's controls via tabbing is critical for our product.
----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/25865025?v=4' /> [Honzasejrak](https://github.com/Honzasejrak) commented on Feb 27, 2017

sirra23

----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/18076683?v=4' /> [webcast33](https://github.com/webcast33) commented on Dec 12, 2017

I see this issue is now scheduled for player release 7.3. Do you have an ECD for that release? Thank you.
----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/2962823?v=4' /> [fdambrosio](https://github.com/fdambrosio) commented on Jan 29, 2018

This is an important feature, thanks 
----

<img width='32' height='32' src='https://avatars.githubusercontent.com/u/2962823?v=4' /> [fdambrosio](https://github.com/fdambrosio) commented on Apr 3, 2018

can you add the aria-label attribute?  
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute
